File: eolymp/prep4/bracket_balancing.cpp
#include<iostream>
#include<stack>
#include<string>

using namespace std;

stack<char> brackets;

int main(){
    int n;
    string line;
    bool flag = true;
    cin >> n;
    for(string line; getline(cin, line);){
        flag = true;
        while(!brackets.empty()){
            brackets.pop();
        }
        cout << line;
        for(char& character : line){
            if(character == '('){
                brackets.push('(');
            }
            else if(character == '['){
                brackets.push('[');
            }
            else if((character == ')') && (brackets.top() == '(')){
                brackets.pop();
                brackets.pop();
            }
            else if((character == ']') && (brackets.top() == '[')){
                brackets.pop();
                brackets.pop();
            }
            else if((character == ')') && (brackets.top() == '[')){
                flag = false;
                break;
            }
            else if((character == ']') && (brackets.top() == '(')){
                flag = false;
                break;
            }
            if((character == (')' || ']')) && brackets.empty()){
                flag = false;
                break;
            }
        }
        if (flag){
            cout << "YES" << "\n";
        }
        else{
            cout << "NO" << "\n";
        }
    }
    return 0;
}

File: eolymp/prep4/bracket_balancing.py
open_list = ["[","("]
close_list = ["]",")"]
def check(brackets):
    stack = []
    for i in brackets:
        if i in open_list:
            stack.append(i)
        elif i in close_list:
            pos = close_list.index(i)
            if ((len(stack) > 0) and
                (open_list[pos] == stack[len(stack)-1])):
                stack.pop()
            else:
                return "No"
    if len(stack) == 0:
        return "Yes"
    else:
        return "No"

print(check(input()))
